toolchain/check/testdata/impl/no_prelude/impl_cycle.carbon

```
@@ -0,0 +1,2431 @@
+  // CHECK:STDERR:   ({} as Point) as (Point as Z);
+  // CHECK:STDERR:                     ^~~~~~~~~~
+  // CHECK:STDERR: fail_impl_simple_cycle.carbon:[[@LINE-10]]:1: note: while deducing parameters of generic declared here [DeductionGenericHere]
+  // CHECK:STDERR: impl forall [T:! Z] T as Z {}
+  // CHECK:STDERR: ^~~~~~~~~~~~~~~~~~~~~~~~~~~~
+  // CHECK:STDERR:
+  ({} as Point) as (Point as Z);
+}
+
+// --- fail_impl_cycle_one_generic_param.carbon
+library "[[@TEST_NAME]]";
+
+import Core;
+
+interface ComparableWith(T:! type) {}
+
+// This creates a dependency cycle with itself.
+impl forall [U:! type, T:! ComparableWith(U)]
+    U as ComparableWith(T) {}
+
+class C {}
+class D {}
+impl C as ComparableWith(D) {}
+
+fn Compare[T:! type, U:! ComparableWith(T)](t: T, u: U) {}
+
+fn F() {
+  // CHECK:STDERR: fail_impl_cycle_one_generic_param.carbon:[[@LINE+20]]:3: error: cycle found in lookup of interface <TODO: interface name> for type `C` [ImplLookupCycle]
+  // CHECK:STDERR:   Compare({} as C, {} as C);
+  // CHECK:STDERR:   ^~~~~~~~~~~~~~~~~~~~~~~~~
+  // CHECK:STDERR: fail_impl_cycle_one_generic_param.carbon:[[@LINE-13]]:1: note: while deducing parameters of generic declared here [DeductionGenericHere]
+  // CHECK:STDERR: impl forall [U:! type, T:! ComparableWith(U)]
+  // CHECK:STDERR: ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+  // CHECK:STDERR: fail_impl_cycle_one_generic_param.carbon:[[@LINE-9]]:1: note: while deducing parameters of generic declared here [DeductionGenericHere]
+  // CHECK:STDERR: fn Compare[T:! type, U:! ComparableWith(T)](t: T, u: U) {}
+  // CHECK:STDERR: ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+  // CHECK:STDERR:
+  // CHECK:STDERR: fail_impl_cycle_one_generic_param.carbon:[[@LINE+10]]:3: error: cannot implicitly convert from `type` to `ComparableWith(C)` [ImplicitAsConversionFailure]
```

Is this error something that should be suppressed if impl lookup fails due to a loop? Or is it that it keeps looking and could still find a relevant impl?

---

The cycle found means that impl lookup fails, which is diagnosed. This means convert does not produce an argument during deduction that will match the parameter. Then conversion also fails to be able to convert to anything compatible, since impl lookup failed.

Generally impl lookup fails silently, and you just get the type conversion error. But in this case, impl loopup also had an error. I think it's correct to then say the conversion failed (possibly due to the impl lookup loop detection or otherwise).

It does keep looking, yes. One impl finding a cycle does not abandon the whole deduce process.